src/middleware/compression.rs

```diff
@@ -0,0 +1,204 @@
+use futures::future::FutureObj;
+
+use crate::body::Body;
+use crate::{middleware::RequestContext, Middleware, Response};
+use http::header::{HeaderValue, ACCEPT_ENCODING, CONTENT_ENCODING};
+
+use std::io::Read;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Encoding {
+    Gzip,
+    Deflate,
+    Identity,
+}
+
+impl Encoding {
+    fn parse(i: &str) -> Option<Encoding> {
+        use self::Encoding::*;
+        match i {
+            "gzip" => Some(Gzip),
+            "deflate" => Some(Deflate),
+            "identity" => Some(Identity),
+            _ => None,
+        }
+    }
+
+    fn header_value(self) -> HeaderValue {
+        use self::Encoding::*;
+        match self {
+            Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Identity => HeaderValue::from_str("identity").unwrap(),
+        }
+    }
+}
+
+/// Tide's compression middleware.
+/// Will pick preferred supported format from "Accept-Encoding" header
+/// and serialize response accordingly
+pub struct Compression {
+    default: Encoding,
+}
+
+impl Compression {
+    pub fn new() -> Self {
+        Compression {
+            default: Encoding::Gzip,
+        }
+    }
+
+    pub fn with_default(default: Encoding) -> Self {
+        Compression { default }
+    }
+
+    fn preferred_encoding(headers: &http::HeaderMap) -> Option<Encoding> {
+        headers
+            .get_all(ACCEPT_ENCODING)
+            .iter()
+            .filter_map(|hval| hval.to_str().ok())
+            .flat_map(|val| val.split(|c| c == ','))
+            .filter_map(|encoding| Encoding::parse(encoding))
```

Comment:
```suggestion
            .filter_map(Encoding::parse)
```
IIRC, clippy have a warning related to that.